chainermn/datasets/shuffle_datablock.py

```diff
@@ -0,0 +1,219 @@
+                # NOTE: recv_cnt is a <0 value
+                send_counts[(src_rank, dst_rank)] = -recv_cnt
+                diff_length[src_rank] += recv_cnt
+                diff_length[dst_rank] = 0
+            if diff_length[src_rank] == 0:
+                break
+
+    # All the diffs are resolved by communication.
+    assert all(d == 0 for d in diff_length)
+    return send_counts, new_lengths
+
+
+def _exchange_block(comm, data, block, data_lengths, block_lengths):
+    send_counts, new_lengths = _calc_alltoall_send_counts(
+        data_lengths, block_lengths)
+
+    # Basically, send data is selected from the newly-loaded `block`,
+    # but in some cases `block` is an empty array
+    # i.e. when adjusting for force_equal_length
+    if len(block) == 0:
+        local_data = data
+    else:
+        local_data = block
+
+    offset = 0
+    send_buf = [[]] * comm.size
+    for dest_rank in range(comm.size):
+        num_elem = send_counts.get((comm.rank, dest_rank), 0)
+        if num_elem == 0:
+            continue
+
+        send_buf[dest_rank] = local_data[offset:offset + num_elem]
+        while len(send_buf[dest_rank]) < num_elem:
+            # In case of force_equal_length, the process may have to send
+            # more data than `block` has. We need to duplicate some elements
+            # from `block`.
+            send_buf[dest_rank].append(random.choice(block))
+        offset = (offset + num_elem) % len(local_data)
+    data += _flatten1(comm.mpi_comm.alltoall(send_buf))
+    return new_lengths
+
+
+def shuffle_data_blocks(comm, blocks, block_size, force_equal_length=True):
+    """Exchange unbalanced blocks of data between all processes
+
+    This function is useful when `scatter_dataset` is not suitable.
+    For instance, the data is huge or the total length is unknown.
+
+    shuffle_datablocks() function works in a iterative way.
+    In a single iteartion, `block_size` items are loaded from `data_blocks`
+    data source and exchanged between distributed processes so the number of
+    elements in each process is roughly equal.
+
+    If `force_equal_length` is True, some elements are duplicated when
+    necessary so all processes have exactly the same number of elements
+    after all iterations.
+
+    :param comm: :class: `~chainermn.communicators.MpiCommunicatorBase`
+    :param blocks: An iterative object to read data blocks.
+    :param force_equal_length: Whether data length of each process is adjusted
+                               by copying some elements, so that all processes
+                               have exactly the same length of data.
+                               This is required in training for correct
+                               iteration/epoch counting.
+                               In evaluation, however, the option can be False
+                               if you don't want duplicated elements.
+    :param block_size: Number of elements read from `data_blocks`
```

Comment:
How to design the scale of `block_size` have to be clearly described with its performance behaviour and what limitation MPI has. For example in `alltoall` it involves several limitations like number of max transfer size by `MPI_alltoall` , limit of pickle size and so on.